<a href="https://colab.research.google.com/github/NEGU93/keras-tcn/blob/master/af_challenge.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install tensorflow==2.12
!pip install keras-tcn

In [2]:
import numpy as np
import pandas as pd
from pathlib import Path
import plotly.express as px
from datetime import datetime, timedelta
from sklearn.preprocessing import StandardScaler
import tensorflow as tf
from tcn import TCN

# Data Exploration

In [3]:
df = pd.read_csv(Path('/content/lounge_attendance.csv'))
df.date = pd.to_datetime(df.date)
df

date lounge_name  total_guests
0    2017-01-01          EK          1295
1    2017-01-01          EL          1294
2    2017-01-01          EM           738
3    2017-01-01           F           869
4    2017-01-01           G           128
...         ...         ...           ...
5442 2019-12-29          EK          1527
5443 2019-12-29          EL          1494
5444 2019-12-29          EM          1018
5445 2019-12-29           F          2042
5446 2019-12-29           G           243

[5447 rows x 3 columns]

In [4]:
df.dtypes

date            datetime64[ns]
lounge_name             object
total_guests             int64
dtype: object

In [5]:
# I usually start plotting the data
fig = px.line(df, x='date', y='total_guests', facet_col='lounge_name', facet_col_wrap=3)
fig.show()

In [6]:
fig = px.box(df, x='lounge_name', y='total_guests',  notched=True, color='lounge_name', points="all")
fig.update_traces(boxmean=True)
fig.show()
# Data seems to be unskewed (mean close to median)

In [7]:
# Note: Data is from 1st of Jan 2017 to almost end of December 2019. So per year I have all days and month, making stats easier to compute seasonaly.


# And getting some stats with the describe feature
df.groupby('lounge_name')['total_guests'].describe()

count         mean         std    min     25%     50%      75%  \
lounge_name                                                                    
EK           1093.0  1707.243367  273.147938  733.0  1524.0  1688.0  1883.00   
EL           1093.0  1432.578225  292.187065  638.0  1218.0  1406.0  1638.00   
EM           1092.0  1022.134615  203.687926  492.0   883.5  1013.0  1158.00   
F            1093.0  2493.444648  591.530569  558.0  2040.0  2523.0  2985.00   
G            1076.0   485.647770  144.527714    4.0   376.0   492.0   599.25   

                max  
lounge_name          
EK           2641.0  
EL           2291.0  
EM           1737.0  
F            3919.0  
G             813.0

In [8]:
# It seems I have missing dates, let's found them
df.date.max() - df.date.min()

Timedelta('1092 days 00:00:00')

In [9]:
def find_missing_dates(df):
    """
    For a given df with columns :code:`lounge_name` and :code:`date`, it finds for every different lounge name, the dates missing, assuming a frequency of one day.
    :param df: Dataframe with columns columns :code:`lounge_name` (str) and :code:`date` (datetime).
    :retruns: dictionary with keys the lounge names and values the missing dates.
    """
    max_date = df.date.max()
    min_date = df.date.min()
    missing_dates_dict = {}

    for lounge_name, group in df.groupby('lounge_name'):
        date_range = pd.date_range(start=min_date, end=max_date)
        missing_dates = date_range.difference(group['date'])
        missing_dates_dict[lounge_name] = missing_dates.tolist()
    return missing_dates_dict

missing_dates = find_missing_dates(df)
missing_dates

{'EK': [],
 'EL': [],
 'EM': [Timestamp('2018-02-08 00:00:00')],
 'F': [],
 'G': [Timestamp('2018-02-08 00:00:00'),
  Timestamp('2019-07-01 00:00:00'),
  Timestamp('2019-07-02 00:00:00'),
  Timestamp('2019-07-03 00:00:00'),
  Timestamp('2019-07-04 00:00:00'),
  Timestamp('2019-07-05 00:00:00'),
  Timestamp('2019-07-06 00:00:00'),
  Timestamp('2019-07-07 00:00:00'),
  Timestamp('2019-07-08 00:00:00'),
  Timestamp('2019-07-09 00:00:00'),
  Timestamp('2019-07-10 00:00:00'),
  Timestamp('2019-07-11 00:00:00'),
  Timestamp('2019-07-12 00:00:00'),
  Timestamp('2019-07-13 00:00:00'),
  Timestamp('2019-07-14 00:00:00'),
  Timestamp('2019-07-15 00:00:00'),
  Timestamp('2019-07-16 00:00:00')]}

In [10]:
# I want to check week, month and year seasonality
df['day_of_week'] = df.date.dt.day_name()
df['month'] = df.date.dt.month
df['year'] = df.date.dt.year
df.head()

date lounge_name  total_guests day_of_week  month  year
0 2017-01-01          EK          1295      Sunday      1  2017
1 2017-01-01          EL          1294      Sunday      1  2017
2 2017-01-01          EM           738      Sunday      1  2017
3 2017-01-01           F           869      Sunday      1  2017
4 2017-01-01           G           128      Sunday      1  2017

In [11]:
# There seems indeed to be a high difference between weekdays, but I would rather see the percentaje to have an idea
df.groupby(['lounge_name', 'day_of_week'])['total_guests'].describe()

count         mean         std     min      25%  \
lounge_name day_of_week                                                    
EK          Friday       156.0  1633.493590  206.420101  1069.0  1509.75   
            Monday       156.0  1952.596154  286.604882  1155.0  1782.75   
            Saturday     156.0  1708.288462  222.770300  1087.0  1590.75   
            Sunday       157.0  1879.891720  265.366186  1062.0  1749.00   
            Thursday     156.0  1539.551282  206.627428   781.0  1414.25   
            Tuesday      156.0  1627.275641  237.369769   733.0  1493.00   
            Wednesday    156.0  1608.500000  201.350521   927.0  1479.75   
EL          Friday       156.0  1337.230769  228.045011   759.0  1169.75   
            Monday       156.0  1713.442308  278.320730   779.0  1578.50   
            Saturday     156.0  1452.910256  252.810228   755.0  1256.75   
            Sunday       157.0  1647.280255  257.276137   803.0  1501.00   
            Thursday     156.0  1251.557692  221.086970   638.0  1086.75   
            Tuesday      156.0  1342.717949  233.378740   669.0  1199.25   
            Wednesday    156.0  1281.532051  195.972724   758.0  1152.25   
EM          Friday       156.0   980.089744  167.299283   578.0   865.25   
            Monday       156.0  1186.134615  175.846197   745.0  1067.25   
            Saturday     156.0  1069.493590  136.371717   712.0   979.50   
            Sunday       157.0  1194.343949  178.796465   684.0  1070.00   
            Thursday     155.0   880.025806  150.518970   492.0   773.00   
            Tuesday      156.0   936.500000  165.609393   556.0   818.75   
            Wednesday    156.0   906.339744  173.762259   537.0   788.00   
F           Friday       156.0  2877.826923  555.965248  1603.0  2525.25   
            Monday       156.0  2748.557692  615.374905  1112.0  2362.75   
            Saturday     156.0  1980.448718  362.556234   749.0  1782.50   
            Sunday       157.0  2055.000000  350.992549   869.0  1892.00   
            Thursday     156.0  2753.519231  508.363884  1489.0  2401.00   
            Tuesday      156.0  2465.365385  527.774195   558.0  2198.25   
            Wednesday    156.0  2576.205128  485.196903  1254.0  2309.25   
G           Friday       154.0   568.266234  132.038222   199.0   493.25   
            Monday       153.0   586.307190  158.635766   113.0   509.00   
            Saturday     154.0   351.701299   72.983939   143.0   305.25   
            Sunday       155.0   378.483871   82.173506   128.0   327.00   
            Thursday     153.0   516.464052  118.350340   142.0   449.00   
            Tuesday      153.0   497.660131  124.134433   112.0   451.00   
            Wednesday    154.0   502.279221  122.121203     4.0   455.00   

                            50%      75%     max  
lounge_name day_of_week                           
EK          Friday       1642.5  1763.50  2200.0  
            Monday       1948.5  2143.50  2641.0  
            Saturday     1699.5  1862.25  2305.0  
            Sunday       1903.0  2084.00  2545.0  
            Thursday     1529.0  1657.25  2122.0  
            Tuesday      1641.0  1799.25  2155.0  
            Wednesday    1620.0  1735.25  2148.0  
EL          Friday       1325.5  1490.25  2074.0  
            Monday       1741.5  1908.50  2219.0  
            Saturday     1431.5  1645.00  2097.0  
            Sunday       1662.0  1822.00  2259.0  
            Thursday     1222.0  1398.75  1910.0  
            Tuesday      1341.5  1498.75  2291.0  
            Wednesday    1262.5  1409.50  1752.0  
EM          Friday        984.0  1090.00  1540.0  
            Monday       1181.5  1324.25  1603.0  
            Saturday     1064.0  1163.00  1400.0  
            Sunday       1194.0  1324.00  1684.0  
            Thursday      896.0   974.00  1315.0  
            Tuesday       952.0  1050.50  1342.0  
            Wednesday     895.0  1008.00  1737.0  
F           Friday       3039.0  3304.5

In [12]:
# Instead of showing the pandas table percentages, let's do some plotly showoff
fig = px.sunburst(
    df,
    path=['lounge_name', 'day_of_week'],
    values='total_guests',
)
fig.update_traces(textinfo="label+percent parent")
fig.show()
# Conlcusions from the graph:
# - Monday has normally higher afluence (and high if not)
# - Thursday lower affluence, except for F which is quite high.

In [13]:
# Let's check monthly seasonality, with another graph to show I am resourceful.
fig = px.histogram(df, x='month', color='lounge_name', y='total_guests', barmode='group')
fig.show()
# Conclusions from this graph: There are peaks around October and another around June.

In [14]:
fig = px.histogram(df, x='year', color='lounge_name', y='total_guests', barmode='group')
fig.show()
# TIP: F lounge makes it hard to see the effect for G for example, but you can show only G with plotly so you will the tendence.
# Per year, there is a tendency to increase afluence, except for lounge G. The weeks on July the lounge was "closed" might have something to do with this. July is a high afluence month.

In [15]:
# Create lagged features for previous date, day of week, and month
df['previous_date_total_guests'] = df.groupby('lounge_name')['total_guests'].shift(1)
df['previous_day_of_week_total_guests'] = df.groupby(['lounge_name', 'day_of_week'])['total_guests'].shift(1)
df['previous_month_total_guests'] = df.groupby(['lounge_name', 'month'])['total_guests'].shift(1)

# Calculate correlations
correlation_previous_date = df['total_guests'].corr(df['previous_date_total_guests'])
correlation_previous_day_of_week = df['total_guests'].corr(df['previous_day_of_week_total_guests'])
correlation_previous_month = df['total_guests'].corr(df['previous_month_total_guests'])

print(f"Correlation with previous date's total guests: {correlation_previous_date}")
print(f"Correlation with previous day of week's total guests: {correlation_previous_day_of_week}")
print(f"Correlation with previous month's total guests: {correlation_previous_month}")

Correlation with previous date's total guests: 0.911761289759513
Correlation with previous day of week's total guests: 0.9482434342306539
Correlation with previous month's total guests: 0.9082786915821456


## Data exploration conclusions
- Data is from 1st of Jan 2017 to almost end of December 2019.
- Seems to be unskewed.
- I have missing dates for lounge EM (only one) and G (about a week, probably the lounge was closed).
  - Did this generated more affluence to other lounges?
- There seems to be a strong seasonality, per month and day of the week, as well as some tendency to increase per year.

# Pre-processing

Objectives:
- Deal with missing values
- Get train-test set

In [16]:
# Create a copy of the DataFrame to avoid modifying the original
df_filled = df.copy()

# Iterate through each lounge and its missing dates
for lounge, missing_dates_list in missing_dates.items():
    for missing_date in missing_dates_list:
        # Find the previous week's date
        previous_week_date = missing_date - pd.Timedelta(days=7)

        # Find the total_guests for the previous week's date for the current lounge
        previous_week_guests = df_filled[
            (df_filled['lounge_name'] == lounge) & (df_filled['date'] == previous_week_date)
        ]['total_guests']

        # If the previous week's data exists, fill the missing date with that value
        if not previous_week_guests.empty:
            new_row = pd.DataFrame({
                'lounge_name': [lounge],
                'date': [missing_date],
                'total_guests': [previous_week_guests.iloc[0]],
                'day_of_week': [missing_date.day_name()],
                'month': [missing_date.month],
                'year': [missing_date.year]
            })
            df_filled = pd.concat([df_filled, new_row], ignore_index=True)

# Sort the DataFrame by date
df_filled = df_filled.sort_values(by='date').reset_index(drop=True)

# Recalculate lagged features (optional, but recommended)
df_filled['previous_date_total_guests'] = df_filled.groupby('lounge_name')['total_guests'].shift(1)
df_filled['previous_day_of_week_total_guests'] = df_filled.groupby(['lounge_name', 'day_of_week'])['total_guests'].shift(1)
df_filled['previous_month_total_guests'] = df_filled.groupby(['lounge_name', 'month'])['total_guests'].shift(1)

In [17]:
# Verify my code worked :)
df_filled[
    (df_filled['lounge_name'] == 'G') &
    (df_filled['date'] > datetime(year=2019, month=6, day=23)) &
    (df_filled['date'] < datetime(year=2019, month=7, day=20))
  ]

date lounge_name  total_guests day_of_week  month  year  \
4524 2019-06-24           G           761      Monday      6  2019   
4525 2019-06-25           G           603     Tuesday      6  2019   
4534 2019-06-26           G           608   Wednesday      6  2019   
4535 2019-06-27           G           695    Thursday      6  2019   
4543 2019-06-28           G           645      Friday      6  2019   
4549 2019-06-29           G           390    Saturday      6  2019   
4550 2019-06-30           G           418      Sunday      6  2019   
4559 2019-07-01           G           761      Monday      7  2019   
4560 2019-07-02           G           603     Tuesday      7  2019   
4568 2019-07-03           G           608   Wednesday      7  2019   
4570 2019-07-04           G           695    Thursday      7  2019   
4578 2019-07-05           G           645      Friday      7  2019   
4580 2019-07-06           G           390    Saturday      7  2019   
4587 2019-07-07           G           418      Sunday      7  2019   
4593 2019-07-08           G           761      Monday      7  2019   
4595 2019-07-09           G           603     Tuesday      7  2019   
4603 2019-07-10           G           608   Wednesday      7  2019   
4605 2019-07-11           G           695    Thursday      7  2019   
4611 2019-07-12           G           645      Friday      7  2019   
4619 2019-07-13           G           390    Saturday      7  2019   
4621 2019-07-14           G           418      Sunday      7  2019   
4625 2019-07-15           G           761      Monday      7  2019   
4630 2019-07-16           G           603     Tuesday      7  2019   
4639 2019-07-17           G           455   Wednesday      7  2019   
4644 2019-07-18           G           513    Thursday      7  2019   
4648 2019-07-19           G           562      Friday      7  2019   

      previous_date_total_guests  previous_day_of_week_total_guests  \
4524                       510.0                              593.0   
4525                       761.0                              498.0   
4534                       603.0                              545.0   
4535                       608.0                              519.0   
4543                       695.0                              579.0   
4549                       645.0                              446.0   
4550                       390.0                              510.0   
4559                       418.0                              761.0   
4560                       761.0                              603.0   
4568                       603.0                              608.0   
4570                       608.0                              695.0   
4578                       695.0                              645.0   
4580                       645.0                              390.0   
4587                       390.0                              418.0   
4593                       418.0                              761.0   
4595                       761.0                              603.0   
4603                       603.0                              608.0   
4605                       608.0                              695.0   
4611                       695.0                              645.0   
4619                       645.0                              390.0   
4621                       390.0                              418.0   
4625                       418.0                              761.0   
4630                       761.0                              603.0   
4639                       603.0                              608.0   
4644                       455.0                              695.0   
4648                       513.0                              645.0   

      previous_month_total_guests  
4524                        510.0  
4525                        761.0  
4534                        603.0  
4535                        608.0  
4543                  

In [18]:
# Another verification
df_filled.groupby('lounge_name')['total_guests'].count()

lounge_name
EK    1093
EL    1093
EM    1093
F     1093
G     1093
Name: total_guests, dtype: int64

In [19]:
# Can't stop testing my code worked.
# Here I just verify all columns in df are also in df_filled... So that I kept all values.
for index, row in df.iterrows():
    # Find the corresponding row in df_filled
    matching_rows = df_filled[
        (df_filled['lounge_name'] == row['lounge_name']) &
        (df_filled['date'] == row['date'])
    ]
    # Check if a matching row exists
    if not matching_rows.empty:
        # Check if all values in the matching row are equal to the original row
        matching_row = matching_rows.iloc[0]
        if not all(row[col] == matching_row[col] for col in ['date', 'total_guests', 'lounge_name']):
            print(f"Mismatch found for row in df:\n{row}\nMatching row in df_filled:\n{matching_row}")
    else:
        print(f"Row in df not found in df_filled:\n{row}")

In [20]:
# I think using 2019-06-30 as train - test split will be Ok, therefore I prevent training on new generated data.
def train_test_split(df_filled, split_date: datetime):
    """
    Splits a dataframe into train and validation using a date as a split.
    It also scales the dataframes and returns the scaler used for that.

    .. warning::

      Data verification like the value of split_date is between the df dates are not done.

    :param df: Dataframe to be splitted with columns :code:`date` and :code:`total_guests`.
    :param split_date: Date to split the dataframe.
    :returns: numpy arrays of train and validation data. With shapes (number of lounges, timeseries of total guests, 1).

      - :code:`train_np`: numpy array with dates before :code:`split_date`
      - :code:`val_np`: numpy array with dates after :code:`split_date`
      - :code:`all_data_np`: numpy array with all the data. second dimension should be the sum of :code:`train_np` and :code:`val_np` second dimension.
      - :code:`scalers`: list of scalers used to scale the data.
    """
    train_np = []
    val_np = []
    all_data_np = []
    scalers = []
    for lounge in df_filled['lounge_name'].unique():
        lounge_df = df_filled[df_filled['lounge_name'] == lounge]
        scaler = StandardScaler()
        # Fit my scaler with all the data
        all_data_np.append(scaler.fit_transform(np.expand_dims(lounge_df['total_guests'].to_numpy(), axis=-1)))
        train_np.append(scaler.transform(np.expand_dims(lounge_df[lounge_df['date'] <= split_date]['total_guests'].to_numpy(), axis=-1)))
        val_np.append(scaler.transform(np.expand_dims(lounge_df[lounge_df['date'] > split_date]['total_guests'].to_numpy(), axis=-1)))
        scalers.append(scaler)   # I save the scaler to un-transform the predictions.

    val_np = np.array(val_np)
    all_data_np = np.array(all_data_np)
    train_np = np.array(train_np)
    return train_np, val_np, all_data_np, scalers


split_date=datetime(year=2019, month=6, day=30)
train_np, val_np, all_data_np, scalers = train_test_split(df_filled, split_date=split_date)
print(train_np.shape)
print(val_np.shape)
print(all_data_np.shape)

(5, 911, 1)
(5, 182, 1)
(5, 1093, 1)


In [21]:
scalers[0].mean_

array([1707.24336688])

# Train

## Model ideas

- I know SARIMA-base models deal well with seasonal and trend timeseries. It could be a strong method to substract information about this effect, and would also offer some data explicability (we could conclude the trend from year to year for excample).
- Meta [Prophet](https://facebook.github.io/prophet/). seems to be quite tailored for this task. Here their own description: *"Prophet is a procedure for forecasting time series data based on an additive model where non-linear trends are fit with yearly, weekly, and daily seasonality, plus holiday effects. It works best with time series that have strong seasonal effects and several seasons of historical data. Prophet is robust to missing data and shifts in the trend, and typically handles outliers well."*.
- Just some regression with Random Forest, GBoost, etc.
- Neural Networks: LSTMs or TCN algorithms (or Mamba / Transformers).

Prophet seems quite an interesting choice, but I have never tried it or used it, so I fear loosing too much time.
I decided to use TCN because it has some complexity (I allows me to showoff, using tf or Pytorch instead of sklearn), it might be a little overkill (but less overkill than Transformers).
In my experience, TCN works similarly in performance than LSTM but are quite easier to train (normally works without any parameter tuning).

I would like to first construct a benchmark to compare my result. I used a "Last day of the week" prediction because it has the higher correlation.

## Training method

I have 3 options for training the TCN:"

1. (BAD) Make model predict D-day + 14.
2. Iterative prediction (predict D-day+1 and use it iteratively to predict all 14 days).
3. Predict 14 days at once.

I choose option 3.

In [22]:
input_layer = tf.keras.Input(shape=(None, 1))
tcn_block = TCN(
    nb_filters=64,
    kernel_size=3,
    nb_stacks=1,
    dilations=(1, 2, 4, 8, 16, 32),
    padding='causal',
    use_skip_connections=True,
    dropout_rate=0.0,
    return_sequences=True,
    activation='relu',
    kernel_initializer='he_normal',
    use_batch_norm=False,
    use_layer_norm=False,
    use_weight_norm=False,
    go_backwards=False,
    return_state=False,
)(input_layer)
dense = tf.keras.layers.Dense(14)(tcn_block)
output_layer = tf.keras.layers.Activation('linear')(dense)
tcn_model = tf.keras.Model(inputs=input_layer, outputs=output_layer)

tcn_model.compile(optimizer='adam', loss='mse')
tcn_model.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, None, 1)]         0         
                                                                 
 tcn (TCN)                   (None, None, 64)          136256    
                                                                 
 dense (Dense)               (None, None, 14)          910       
                                                                 
 activation (Activation)     (None, None, 14)          0         
                                                                 
Total params: 137,166
Trainable params: 137,166
Non-trainable params: 0
_________________________________________________________________


In [23]:
def get_intput_and_target(array_np, forecast_total: int = 14):
    """
    Gets input and target of model for a predictive task.
    :param array_np: Numpy array with shape (batch, features, 1).
    :returns:

      - :code:`x` same as numpy array without the last forecast_total samples
      - :code:`y` numpy array with shape (batch, features, forecast). Where last dimension corresponds to a different shift, starting from 1.
    """
    x = array_np[:, :-14]
    y = np.zeros(shape=x.shape[:-1] + (14,))

    for forecast_time in range(13):
        y[:,:,forecast_time] = array_np[:, (forecast_time+1):(-14+forecast_time+1)].squeeze()
    y[:,:,-1] = array_np[:, 14:].squeeze()
    return x, y

# The model, for sample i, needs to predict i+N (N being the shift y of last dimension)
x_train, y_train = get_intput_and_target(train_np)
x_val, y_val = get_intput_and_target(val_np)
x_train.shape

(5, 897, 1)

In [24]:
epochs=500
history = tcn_model.fit(epochs=epochs, x=x_train, y=y_train, validation_data=(x_val, y_val), verbose=0)

In [25]:
train_history = history.history
train_history.update({'epochs': list(range(epochs))})
fig = px.line(
    pd.DataFrame(
        train_history
    ),
    x='epochs',
    y=['loss', 'val_loss']
)
fig.show()

In [26]:
prediction = tcn_model.predict(all_data_np)
unnormalized_prediction = []
for i, scale in enumerate(scalers):
    unnormalized_prediction.append(scale.inverse_transform(prediction[i]))
unnormalized_prediction = np.ceil(np.array(unnormalized_prediction)).astype(int)  # Scale up because better to miss on more

1/1 [==============================] - 1s 563ms/step


In [27]:
unnormalized_prediction.shape

(5, 1093, 14)

In [28]:
tcn_model.predict(all_data_np).shape

1/1 [==============================] - 0s 77ms/step


(5, 1093, 14)

In [29]:
for forecast_time in range(1, 15):
  df_filled[f'prediction_{forecast_time}'] = np.nan
  for i, lounge in enumerate(df_filled['lounge_name'].unique()):
      mask = df_filled['lounge_name'] == lounge
      indices = df_filled.loc[mask].iloc[forecast_time:].index
      df_filled.loc[indices, f'prediction_{forecast_time}'] = unnormalized_prediction[i, :-forecast_time, forecast_time-1]
df_filled[df_filled['lounge_name'] == lounge].head(14)

date lounge_name  total_guests day_of_week  month  year  \
4  2017-01-01           G           128      Sunday      1  2017   
9  2017-01-02           G           205      Monday      1  2017   
10 2017-01-03           G           218     Tuesday      1  2017   
19 2017-01-04           G           266   Wednesday      1  2017   
20 2017-01-05           G           300    Thursday      1  2017   
29 2017-01-06           G           204      Friday      1  2017   
30 2017-01-07           G           200    Saturday      1  2017   
39 2017-01-08           G           341      Sunday      1  2017   
40 2017-01-09           G           537      Monday      1  2017   
49 2017-01-10           G           484     Tuesday      1  2017   
51 2017-01-11           G           374   Wednesday      1  2017   
59 2017-01-12           G           437    Thursday      1  2017   
60 2017-01-13           G           467      Friday      1  2017   
69 2017-01-14           G           252    Saturday      1  2017   

    previous_date_total_guests  previous_day_of_week_total_guests  \
4                          NaN                                NaN   
9                        128.0                                NaN   
10                       205.0                                NaN   
19                       218.0                                NaN   
20                       266.0                                NaN   
29                       300.0                                NaN   
30                       204.0                                NaN   
39                       200.0                              128.0   
40                       341.0                              205.0   
49                       537.0                              218.0   
51                       484.0                              266.0   
59                       374.0                              300.0   
60                       437.0                              204.0   
69                       467.0                              200.0   

    previous_month_total_guests  prediction_1  ...  prediction_5  \
4                           NaN           NaN  ...           NaN   
9                         128.0         141.0  ...           NaN   
10                        205.0         302.0  ...           NaN   
19                        218.0         361.0  ...           NaN   
20                        266.0         388.0  ...           NaN   
29                        300.0         268.0  ...         392.0   
30                        204.0         219.0  ...         312.0   
39                        200.0         221.0  ...         194.0   
40                        341.0         348.0  ...         371.0   
49                        537.0         426.0  ...         282.0   
51                        484.0         394.0  ...         317.0   
59                        374.0         351.0  ...         309.0   
60                        437.0         399.0  ...         380.0   
69                        467.0         433.0  ...         324.0   

    prediction_6  prediction_7  prediction_8  prediction_9  prediction_10  \
4            NaN           NaN           NaN           NaN            NaN   
9            NaN           NaN           NaN           NaN            NaN   
10           NaN           NaN           NaN           NaN            NaN   
19           NaN           NaN           NaN           NaN            NaN   
20           NaN           NaN           NaN           NaN            NaN   
29           NaN           NaN           NaN           NaN            NaN   
30         507.0           NaN           NaN           NaN            NaN   
39         276.0         409.0           NaN           NaN            NaN   
40         358.0         448.0         293.0           NaN            NaN   
49         433.0         250.0         375.0         388.0            NaN   
51         260.0         312.0         303.0         307.0          405.0   
59     

In [30]:
# import plotly.graph_objects as go

fig = px.line(
    df_filled,
    x='date',
    y=['total_guests', 'prediction_1', 'prediction_14'],
    facet_col='lounge_name',
    facet_col_wrap=3
)
fig.add_vline(x=split_date.timestamp() * 1000, annotation_text="test")
fig.show()

# Result Analysis

In [31]:
errors = []
for forecast_time in range(1, 15):
  df_filled[f'error_{forecast_time}'] = df_filled['total_guests'] - df_filled[f'prediction_{forecast_time}']
  errors.append(f'error_{forecast_time}')
df_filled['error_benchmark'] = df_filled['total_guests'] - df_filled['previous_day_of_week_total_guests']
errors.append('error_benchmark')

In [32]:
df_filled[errors].describe()
# Conclusions:
# - STD increases with forecast time (quite expected)
# - Bias to predict less than actually happens. This is bad, I would like actually the contrary.
#   - Can just add an offset, but it would be classy to change the loss function to penalize this.
# - 50% of the time I am under 100 people error.

error_1      error_2      error_3      error_4      error_5  \
count  5460.000000  5455.000000  5450.000000  5445.000000  5440.000000   
mean      2.988828     7.472777     3.578349     2.923783     8.873346   
std     209.087385   225.678502   218.494412   220.090882   230.961028   
min   -1525.000000 -1588.000000 -1573.000000 -1534.000000 -1784.000000   
25%    -101.000000  -104.000000  -107.000000  -102.000000  -107.000000   
50%       5.000000     9.000000     4.500000     3.000000     9.000000   
75%     113.000000   120.000000   118.000000   114.000000   125.000000   
max    1203.000000  1155.000000  1015.000000  1482.000000  1223.000000   

           error_6      error_7      error_8      error_9     error_10  \
count  5435.000000  5430.000000  5425.000000  5420.000000  5415.000000   
mean      6.311868     4.023389     3.274286     9.354244     7.158079   
std     231.918241   237.315802   234.662744   252.825076   248.026913   
min   -1476.000000 -1480.000000 -1321.000000 -1536.000000 -1679.000000   
25%    -108.000000  -111.000000  -111.000000  -110.000000  -112.000000   
50%      11.000000     8.500000     5.000000    13.000000    12.000000   
75%     125.000000   128.000000   124.000000   135.000000   134.500000   
max    1132.000000  1323.000000  1467.000000  1474.000000  1237.000000   

          error_11     error_12     error_13     error_14  error_benchmark  
count  5410.000000  5405.000000  5400.000000  5395.000000      5430.000000  
mean      7.300739     5.608326     5.344074     5.979240         0.952302  
std     252.826960   251.672076   249.394750   249.191813       241.980437  
min   -1535.000000 -1592.000000 -1844.000000 -1517.000000     -1724.000000  
25%    -113.000000  -119.000000  -110.250000  -112.000000      -111.000000  
50%      13.000000    12.000000    12.000000    16.000000         1.000000  
75%     136.000000   136.000000   131.000000   129.000000       113.000000  
max    1481.000000  1285.000000  1077.000000  1283.000000      1635.000000

In [33]:
df_filled[df_filled['lounge_name'] == 'G'][errors].describe()

error_1      error_2      error_3      error_4      error_5  \
count  1092.000000  1091.000000  1090.000000  1089.000000  1088.000000   
mean      2.846154    -4.720440     0.722936    -2.549128     1.934743   
std      90.122903    99.927633   101.158289   104.280296   111.170284   
min    -429.000000  -502.000000  -524.000000  -481.000000  -532.000000   
25%     -52.000000   -58.500000   -57.750000   -58.000000   -61.000000   
50%       5.500000     0.000000     5.000000    -1.000000    13.000000   
75%      59.000000    55.000000    66.000000    66.000000    72.000000   
max     347.000000   324.000000   347.000000   309.000000   330.000000   

           error_6      error_7      error_8      error_9     error_10  \
count  1087.000000  1086.000000  1085.000000  1084.000000  1083.000000   
mean     10.174793    -3.532228    -0.022120    -0.576568     0.228070   
std     108.610913   114.270132   109.012929   118.264289   114.712131   
min    -452.000000  -466.000000  -465.000000  -620.000000  -429.000000   
25%     -51.000000   -66.000000   -62.000000   -66.250000   -64.500000   
50%      16.000000     2.000000     6.000000     8.500000    12.000000   
75%      83.000000    69.000000    69.000000    79.000000    76.000000   
max     338.000000   330.000000   314.000000   329.000000   307.000000   

          error_11     error_12     error_13     error_14  error_benchmark  
count  1082.000000  1081.000000  1080.000000  1079.000000      1086.000000  
mean     -3.449168     6.782609    -3.337963     2.018536        -0.064457  
std     118.775941   118.045518   118.528801   119.582831       121.612929  
min    -603.000000  -609.000000  -547.000000  -519.000000      -517.000000  
25%     -63.000000   -60.000000   -65.000000   -62.000000       -57.750000  
50%       7.000000    16.000000     6.500000    21.000000         2.000000  
75%      71.750000    85.000000    78.250000    82.000000        66.000000  
max     373.000000   311.000000   328.000000   366.000000       481.000000

In [34]:
px.box(df_filled, x='lounge_name', y=['error_1'],  notched=True, color='lounge_name', points="all")

In [35]:
fig = px.line(
    df_filled,
    x='date',
    y=['error_1', 'error_benchmark'],
    facet_col='lounge_name',
    facet_col_wrap=3
)
fig.add_vline(x=split_date.timestamp() * 1000, annotation_text="test")
fig.show()

In [36]:
fig = px.histogram(df_filled, x='error_1', color='lounge_name', barmode='group')
fig.show()

# Create prediction and save

In [37]:
x, y = get_intput_and_target(all_data_np)                      # Retrain now with full dataset
history = tcn_model.fit(x=x, y=y, epochs=5000, verbose=0)      # Why not re-train on previous one?
prediction = tcn_model.predict(all_data_np)

1/1 [==============================] - 0s 79ms/step


In [38]:
result_df = pd.DataFrame(columns=['date', 'lounge_name', 'total_guests'])
for i, lounge in enumerate(df_filled['lounge_name'].unique()):
    # import pdb; pdb.set_trace()
    df = pd.DataFrame({
        'date': pd.date_range(start="2019-12-30", end="2020-01-12"),
        'total_guests': np.ceil(scalers[i].inverse_transform(prediction[i])[-1, :]).astype(int),
        'lounge_name': [lounge,] * 14,
    })
    result_df = pd.concat([result_df, df])
result_df.sort_values(['date', 'lounge_name'], inplace=True, ignore_index=True)
result_df

<ipython-input-38-5d14ec98b8f8>:9: FutureWarning:

The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.



date lounge_name total_guests
0  2019-12-30          EK         1703
1  2019-12-30          EL         2045
2  2019-12-30          EM          986
3  2019-12-30           F         2194
4  2019-12-30           G          272
..        ...         ...          ...
65 2020-01-12          EK         1774
66 2020-01-12          EL         2796
67 2020-01-12          EM         1095
68 2020-01-12           F         2081
69 2020-01-12           G          288

[70 rows x 3 columns]

In [39]:
result_df.to_csv('/content/lounge_attendance_predictions.csv', index=False)

In [ ]:
df1 = pd.read_csv(Path('/content/lounge_attendance.csv'))
df1['type'] = 'data'
df2 = pd.read_csv(Path('/content/lounge_attendance_predictions.csv'))
df2['type'] = 'prediction'

fig = px.line(
    pd.concat([df1, df2]),
    x='date',
    y='total_guests',
    color='type',
    facet_col='lounge_name',
    facet_col_wrap=3
)
fig.show()

# Conclusion

## Data Exploration

- Data had missing dates
- High seasonality and trend

## Model

Nothing was done for the fact that 'favor a modelization that would predict slightly more guests than there are'.
The only thing was a ceil for the final output. However, some ideas:

- Do a loss function that penalizes more to predict less than more.
- Fit a gauss function on the error curve and create a shifted threshold. Error theory will even let you choose a threshold that a assures with a certain confidence, that you won't have underpredictions.

## Bonus Question 1:

Error theory says that if EK, EL and EM has 1% error, the sum will also have % error. However, rounding errors (I do ceil) will introduce a higher error. i.e. if all EK, EL and EM have 100.3 guests, they will be predicted as 101, E would have 303 with this method but a more accurate prediction would be 301.

## Bonus Question 2:

I will probably use a similar approach as the gauss graph I did before. Take into account to use it only on dates not used for training of course.
It would be better to use percentage instead of guests error, as they are supposed to increase per year if the yearly tendency is indeed to increase.


In [ ]:
!apt-get install texlive texlive-xetex texlive-latex-extra pandoc
!pip install pypandoc

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
!cp 'drive/My Drive/Colab Notebooks/af_challenge.ipynb' ./

In [ ]:
!jupyter nbconvert --to PDF "af_challenge.ipynb"